API/Auth/DlcsBasicAuthenticationHandler.cs

```diff
@@ -0,0 +1,147 @@
+        }
+
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            // skip authentication if endpoint has [AllowAnonymous] attribute
+            // Not all API paths require auth...
+            var endpoint = Context.GetEndpoint();
+            if (endpoint?.Metadata?.GetMetadata<IAllowAnonymous>() != null)
+                return AuthenticateResult.NoResult();
+            
+            // ...but any not marked must have the auth header
+            if (!Request.Headers.ContainsKey("Authorization"))
+                return AuthenticateResult.Fail("Missing Authorization Header in request");
+
+            var headerValue = Request.GetAuthHeaderValue(AuthenticationHeaderUtils.BasicScheme);
+            if (headerValue == null)
+            {
+                return AuthenticateResult.Fail("Missing Authorization Header in request");
+            }
+
+            // for a path like /customers/23/queue, the resourceCustomerId is 23.
+            // This isn't necessarily the customer that owns the api key being used on the call!
+            int? resourceCustomerId = null;
+            if (Request.RouteValues.TryGetValue("customerId", out var customerIdRouteVal))
+            {
+                if (int.TryParse(customerIdRouteVal.ToString(), out int result))
+                {
+                    resourceCustomerId = result;
+                }
+            }
+            var authHeader = Encoding.UTF8.GetString(Convert.FromBase64String(headerValue.Parameter));
+            string[] keyAndSecret = authHeader.Split(':');
+            var apiCaller = new {Key = keyAndSecret[0], Secret = keyAndSecret[1]};
+            var customerForKey = await customerRepository.GetCustomerForKey(apiCaller.Key, resourceCustomerId);
+            if (customerForKey == null)
+            {
+                return AuthenticateResult.Fail("No customer found for this key that is permitted to access this resource");
+            }
+
+            if (apiCaller.Secret != deliveratorApiAuth.GetApiSecret(customerForKey, Options.Salt, apiCaller.Key))
+            {
+                return AuthenticateResult.Fail("Invalid credentials");
+            }
+
+            // We still have some checks to do before authenticating this user.
+            // Some of these could be classed as authorisation rather than authentication, though they are not user-specific.
+            if (resourceCustomerId.HasValue)
+            {
+                // the request is for a particular customer's resource (e.g., an asset)
+                if (customerForKey.Id != resourceCustomerId.Value)
+                {
+                    // ... but the requester is not this customer. Only proceed if they are admin.
+                    if (!customerForKey.Administrator)
+                    {
+                        return AuthenticateResult.Fail("Only admin user may access this customer's resource.");

```

Comment:
Is it possible to end up here? If I'm customer 10 (not admin) and try to access customer 11's resource, wouldn't `customerForKey` be null?